src/object.c

```diff
@@ -617,9 +617,10 @@ void trimStringObjectIfNeeded(robj *o) {
      * 2. When utilizing the argument caching mechanism in Lua.
      * 3. When the function is called from a module that has allocated a larger buffer than needed. */
     if (len >= PROTO_MBULK_BIG_ARG ||
-       (server.executing_client &&
-       (server.executing_client->flags & CLIENT_MODULE ||
-       (server.executing_client->flags & CLIENT_SCRIPT && len < LUA_CMD_OBJCACHE_MAX_LEN)))) {        
+        (server.executing_client &&
+         (server.executing_client->flags & CLIENT_MODULE ||
```

Comment:
so considering the new realization that this code isn't gonna make any impact for modules anyway (due to the refcount), maybe we wanna now remove the mention for modules here?
p.s. this means that we could have used script_caller and not wait for executing client :smile:

Reply:
It will make a difference if the module calls directly `RedisModule_TrimStringAllocation`

Reply:
ohh, right. well, luckily we have a test for that now.

Reply:
hold on... RedisModule_TrimStringAllocation can be called from a module command, and in that case the executing_client doesn't have the MODULE flag (it has it only if the module called RM_Call).
can you please check why the test passes and / or point me to what i'm missing?

Reply:
It pass since the len is > `PROTO_MBULK_BIG_ARG`

Reply:
Oh, your'e talking about the other test,
Yes, looks like your'e right - the test pass only because in test basics we are in a recursive call:
```
#0  trimStringObjectIfNeeded (o=0x7f37b66643c8) at object.c:614
#1  0x0000000000501812 in RM_TrimStringAllocation (str=0x7f37b66643c8) at module.c:2741
#2  0x00007f37acc233c7 in TestTrimString (ctx=0x7fff6c3c59a0, argv=<optimized out>, argc=<optimized out>) at basics.c:419
#3  0x00000000004fe401 in RedisModuleCommandDispatcher (c=0x7f37b6725900) at module.c:833
#4  0x0000000000452429 in call (c=0x7f37b6725900, flags=4) at server.c:3485
#5  0x0000000000507be9 in RM_Call (ctx=0x7fff6c3c5d00, cmdname=0x7f37acc26ed2 "test.string.trim", fmt=0x7f37acc26cc1 "")
    at module.c:6198
#6  0x00007f37acc24b6f in TestBasics (ctx=0x7fff6c3c5d00, argv=<optimized out>, argc=<optimized out>) at basics.c:877
#7  0x00000000004fe401 in RedisModuleCommandDispatcher (c=0x7f37b6725600) at module.c:833
#8  0x0000000000452429 in call (c=0x7f37b6725600, flags=3) at server.c:3485
#9  0x0000000000453e4c in processCommand (c=0x7f37b6725600) at server.c:4096
```

Reply:
so let's make a test that's in a plain module command (not via RM_Call, and drop that condition)

Reply:
@uriyage can you handle the above and we'll merge this PR?

Reply:
@oranagra, If we drop the condition it won't trim the string, what test do we need in that case? that it doesn't trim it? or do you mean test that it does trim if the buffer is > 16KB? 
Also, this change will basically make `RM_TrimStringAllocation` a noop if the buffer is < 16KB, are we OK with that?

Reply:
you realized that the check for CLIENT_MODULE isn't helpful for the test we made (RM_Call with big string from the network), since RM_Call increment the refcount, but luckily the string is trimmed on the way out of RedisModuleCommandDispatcher, so in that sense there's no need for that condition in trimStringObjectIfNeeded.

on the other hand, i suppose we're missing a test that a module command that doesn't use RM_Call is storing a string in the database (either RM_StringSet, or maybe also one that's creating a module data type that holds a RedisModuleString pointer in it) and uses RM_TrimStringAllocation on a string it created itself (which is smaller than PROTO_MBULK_BIG_ARG but would still have excess that should be trimmed)

am i missing something?

Reply:
on a second thought, there's no need for a test that registers a data type (too much code to test). calling RM_StringSet, and later measuring it's memory, or even asserting that RM_TrimStringAllocation does its work is enough.

Reply:
or actually that's already covered by the existing test you added (`TestTrimString`), so maybe we can just drop the CLIENT_MODULE (replacing it with a comment that mentions the refcount and RedisModuleCommandDispatcher) and everything will work as expected?

Reply:
But `TestTrimString` won't pass if we drop the CLIENT_MODULE. we need to remove this test if we are to remove the check for CLIENT_MODULE.

Reply:
ohh right (sorry for my luck of focus).
RM_TrimStringAllocation calls trimStringObjectIfNeeded which although not getting a client pointer, uses the executing_client.

so in that case, what happens if a module calls RM_TrimStringAllocation not via a command (e.g. a module timer)?
maybe we need to add an explicit flag to trimStringObjectIfNeeded? or skip the trimming if executing_client is NULL?
and add a test using a timer or a detached context / thread.

am i missing something again?

Reply:
Yes if a module calls `RM_TrimStringAllocation` not via `RM_CALL` it won't trim it. 
I think adding a arg to `trimStringObjectIfNeeded` is a good idea.
So I will remove the CLIENT_MODULE , add an argument `trim_small_values ` leave the `TestTrimString` and add a test using timer.
LMK if you think otherwise.

Reply:
sounds. good. thanks.

Reply:
@oranagra, done - I realized that since we remove the CLIENT_MODULE there is no need for additional timer test as `RM_TrimStringAllocation` will work the same no matter the flow.